core/src/cluster_info.rs

```diff
@@ -1533,44 +1536,54 @@ impl ClusterInfo {
         }
     }
 
-    /// Splits a Vec of CrdsValues into a nested Vec, trying to make sure that
-    /// each Vec is no larger than `MAX_PROTOCOL_PAYLOAD_SIZE`
+    /// Splits an input feed of serializable data into chunks where the sum of
+    /// serialized size of values within each chunk is no larger than
+    /// max_chunk_size.
     /// Note: some messages cannot be contained within that size so in the worst case this returns
     /// N nested Vecs with 1 item each.
-    fn split_gossip_messages(msgs: Vec<CrdsValue>) -> Vec<Vec<CrdsValue>> {
-        let mut messages = vec![];
-        let mut payload = vec![];
-        let base_size = serialized_size(&payload).expect("Couldn't check size");
-        let max_payload_size = MAX_PROTOCOL_PAYLOAD_SIZE - base_size;
-        let mut payload_size = 0;
-        for msg in msgs {
-            let msg_size = msg.size();
-            // If the message is too big to fit in this batch
-            if payload_size + msg_size > max_payload_size as u64 {
-                // See if it can fit in the next batch
-                if msg_size <= max_payload_size as u64 {
-                    if !payload.is_empty() {
-                        // Flush the  current payload
-                        messages.push(payload);
-                        // Init the next payload
-                        payload = vec![msg];
-                        payload_size = msg_size;
+    fn split_gossip_messages<I, T>(
+        max_chunk_size: usize,
+        data_feed: I,
+    ) -> impl Iterator<Item = Vec<T>>
+    where
+        T: Serialize + Debug,
+        I: IntoIterator<Item = T>,
+    {
+        let mut data_feed = data_feed.into_iter().fuse();
```

Comment:
is the fuse here just a precaution? Do we care about a more stringent guarantee that there are no more items in the iterator after the first `None` (i.e. something like a flag once a None is observed, and an assert in the below loop?)